include/navigation/differencers.h

```diff
@@ -87,28 +94,101 @@ class AbsDiff
+        {
+            // Get pixel-wise absolute difference
+            cv::absdiff(src1, src2, dst);
+            BOB_ASSERT(dst.type() == CV_8UC1);
+            auto dstMat = dst.getMat();
+
+            mask1.combine(mask2, m_CombinedMask);
+            m_CombinedMask.apply(dstMat, dstMat);
+
+            // Square the differences
+            const auto sz = dstMat.size();
+            m_Differences.resize(sz.width * sz.height);
+            const auto sq = [](uint8_t val) {
+                float fval = val;
+                return fval * fval;
+            };
+            std::transform(dstMat.datastart, dstMat.dataend, m_Differences.begin(), sq);
+
+            return m_CombinedMask.countUnmaskedPixels(src1.size());
+        }
+
+        float mean(cv::InputArray &, size_t count, const ImgProc::Mask &,
+                   const ImgProc::Mask &)
+        {
+            return sqrtf(cv::sum(m_Differences)[0] / (float) count);
+        }
 
-    static inline float mean(const float sum, const float n)
+    private:
+        // We need a second scratch variable to store square differences
+        std::vector<float> m_Differences;
+        ImgProc::Mask m_CombinedMask;
+    };
+};
+
+//------------------------------------------------------------------------
+// BoBRobotics::Navigation::CorrCoefficient
+//------------------------------------------------------------------------
+/*!
+ * \brief For calculating difference based on Pearson's correlation coefficient
+ *
+ * Can be passed to PerfectMemory as a template parameter.
+ *
+ * Note that we use 1 - abs(Pearson's rho) to give us a difference value, as we
+ * want to calculate the dissimilarity between images.
+ */
+struct CorrCoefficient {
+    /*
+     * NB: This template parameter is unused, but left in for consistency with
+     * other differencer classes.
+     */
+    template<class VecType = void>
+    class Internal
     {
-        return sqrt(sum / n);
-    }
+    public:
+        float operator()(cv::InputArray &src1, cv::InputArray &src2,
+                         const ImgProc::Mask &mask1 = {},
+                         const ImgProc::Mask &mask2 = {})
+        {
+            // Don't support masks for now
+            BOB_ASSERT(mask1.empty() && mask2.empty());
+
+            const auto allSame = [](cv::InputArray &arr) {
+                const auto m = arr.getMat();
+                BOB_ASSERT(m.type() == CV_8UC1);
+                auto beg = m.datastart;
+                auto end = m.dataend;
+                return std::all_of(beg + 1, end,
+                                   [beg](uint8_t val) { return val == *beg; });
+            };
+            if (allSame(src1) || allSame(src2)) {
+                throw std::invalid_argument("Vectors src1 and src2 must have "
+                                            "more than one unique value (e.g. "
+                                            "they cannot be all zeros)");
+            }
```

Comment:
This check seems a bit heavyweight for something to do every frame. Not really sure what the alternative is though - accepting that if you put garbage in you get garbage out would be my vote

Reply:
This did occur to me too... I think the performance penalty shouldn't be too bad in general though as ``allSame()`` will return false when it hits upon the first non-matching pixel, which is likely to happen fairly early on.

An alternative would be to only do these additional checks in the case that the correlation coefficient comes out as zero (which seems to be what happens in this error case).

I'm a little concerned about relying on us not passing garbage in though as I can easily imagine a case where an all-black frame is passed in by mistake...